CreeDictionary/search_quality/views.py

```diff
@@ -0,0 +1,116 @@
+import json
+import logging
+import time
+from json import dumps
+from statistics import mean
+
+from django.contrib.admin.views.decorators import staff_member_required
+from django.contrib.auth.mixins import LoginRequiredMixin
+from django.http import (
+    Http404,
+    HttpResponse,
+    StreamingHttpResponse,
+    HttpResponseBadRequest,
+)
+from django.urls import reverse
+from django.utils.html import conditional_escape
+from django.utils.text import slugify
+from django.views.generic import TemplateView
+
+from . import RESULTS_DIR, DEFAULT_SAMPLE_FILE
+from .analyze_results import analyze, load_results_file, count_and_annotate_dupes
+from .sample import load_sample_definition, gen_run_sample
+
+logger = logging.getLogger(__name__)
+
+
+class IndexView(LoginRequiredMixin, TemplateView):
+    template_name = "search_quality/index.html"
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context["json_gz"] = (path.name for path in RESULTS_DIR.glob("*.json.gz"))
+        return context
+
+
+class ResultsAnalysisView(LoginRequiredMixin, TemplateView):
+    template_name = "search_quality/results-analysis.html"
+
+    def get(self, request, *args, **kwargs):
+        # ?json=query returns JSON for query
+        raw = request.GET.get("json", None)
+        if raw:
+            file = self._resolve_url_param_to_file(kwargs["path"])
+
+            search_results = load_results_file(file)
+
+            for k, v in search_results.items():
+                if k == raw:
+                    count_and_annotate_dupes(v["results"])
+                    return HttpResponse(
+                        dumps(v, ensure_ascii=False, indent=2),
+                        content_type="application/json",
+                    )
+            raise Http404()
+
+        return super().get(request, *args, **kwargs)
+
+    def _resolve_url_param_to_file(self, file):
+        file = RESULTS_DIR / file
+        if not file.is_file():
+            raise Http404()
+        return file
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+
+        file = self._resolve_url_param_to_file(kwargs["path"])
+
+        context["name"] = file.name
+        analysis = analyze(file, load_sample_definition(DEFAULT_SAMPLE_FILE))
+        context["analysis"] = analysis
+
+        for k in [
+            "topthree_score_percent",
+            "threeten_score_percent",
+            "total_duplicate_count",
+            "time_taken_seconds",
+        ]:
+            context[f"average_{k}"] = mean(r[k] for r in analysis.values())
+
+        return context
+
+
+@staff_member_required
+def run(request):
+    if request.method != "POST":
+        raise Http404()
```

Comment:
Pedantic suggestion: this should respond with status [`405 Method Not Allowed`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405).

Reply:
I went with `@require_http_methods(["POST"])`, which does sent that status. I was a little disappointed though that it sends blank pages to the client, instead of at least a little “405 Method not allowed” in the body.